src/thor/expansion_action.cc

```diff
@@ -0,0 +1,188 @@
+  SetValueByPointer(dom, "/properties/algorithm", "unidirectional_dijkstra");
+  SetValueByPointer(dom, "/type", "FeatureCollection");
+  SetValueByPointer(dom, "/features/0/type", "Feature");
+  SetValueByPointer(dom, "/features/0/geometry/type", "MultiLineString");
+  SetValueByPointer(dom, "/features/0/geometry/coordinates", Value(kArrayType));
+  SetValueByPointer(dom, "/features/0/properties", Value(kArrayType));
+  for (const auto& prop : exp_props_pbf) {
+    rapidjson::Pointer(kPropPaths[static_cast<Options_ExpansionProps>(prop)])
+        .Set(dom, Value(kArrayType));
+    exp_props_set.insert(static_cast<Options_ExpansionProps>(prop));
+  }
+
+  // a lambda that the path algorithm can call to add stuff to the dom
+  // route and isochrone produce different GeoJSON properties
+  auto track_expansion = [&dom, &exp_action, &opp_edges, &gen_factor, &skip_opps,
+                          &exp_props_set](baldr::GraphReader& reader, baldr::GraphId edgeid,
+                                          const char* algorithm = nullptr,
+                                          const char* status = nullptr, const float duration = 0,
+                                          const u_int32_t distance = 0, const float cost = 0) {
+    auto tile = reader.GetGraphTile(edgeid);
+    if (tile == nullptr) {
+      LOG_ERROR("thor_worker_t::expansion error, tile no longer available" +
+                std::to_string(edgeid.Tile_Base()));
+      return;
+    }
+    const auto* edge = tile->directededge(edgeid);
+    // unfortunately we have to call this before checking if we can skip
+    // else the tile could change underneath us when we get the opposing
+    auto shape = tile->edgeinfo(edge).shape();
+
+    // if requested, skip this edge in case its opposite edge has been added
+    // before (i.e. lower cost) else add this edge's id to the lookup container
+    if (skip_opps) {
+      auto opp_edgeid = reader.GetOpposingEdgeId(edgeid, tile);
+      if (opp_edgeid && opp_edges.count(opp_edgeid))
+        return;
+      opp_edges.insert(edgeid);
+    }
+
+    if (!edge->forward())
+      std::reverse(shape.begin(), shape.end());
+    Polyline2<PointLL>::Generalize(shape, gen_factor, {}, false);
+
+    // make the geom
+    auto& a = dom.GetAllocator();
+    auto* coords = GetValueByPointer(dom, "/features/0/geometry/coordinates");
+    coords->GetArray().PushBack(Value(kArrayType), a);
+    auto& linestring = (*coords)[coords->Size() - 1];
+    for (const auto& p : shape) {
+      linestring.GetArray().PushBack(Value(kArrayType), a);
+      auto point = linestring[linestring.Size() - 1].GetArray();
+      point.PushBack(p.first, a);
+      point.PushBack(p.second, a);
+    }
+
+    // no properties asked for, don't collect any
+    if (!exp_props_set.size()) {
+      return;
+    }
+
+    // make the properties
+    if (algorithm)
+      SetValueByPointer(dom, "/properties/algorithm", algorithm);
```

Comment:
we default the algorithm attribute to dijkstra, so isochrones don't have to set it, but others will